core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala

```diff
@@ -460,7 +460,9 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
         |                              on one of the worker machines inside the cluster ("cluster")
         |                              (Default: client).
         |  --class CLASS_NAME          Your application's main class (for Java / Scala apps).
-        |  --name NAME                 A name of your application.
+        |  --name NAME                 A name of your application. In yarn-cluster mode the name
```

Comment:
@vanzin `SparkConf.set("spark.app.name", "foo")` will not work in cluster-mode applications, since this line of code will run when the AM ignites the main user class. This setting will not take effect, since it will run on YARN, while the client already set the application name during application-submission. (And you are not able to change the app's name later on with YARN.)

I suggest to explicitly emphasize this behavior and limitation at `setAppName`, and I would encourage YARN users to use `spark-submit --name` while deploying. Maybe we should deprecate `setAppName`, because it can take no effect in cluster-mode.